### PR TITLE
re-publish and re-discover immediately when the node disconnects

### PR DESCRIPTION
**Pull request by AmnaSnene**

### CODE REVIEW

File: src/dhtrunner.cpp [hunk 1]
```diff
@@ -690,8 +696,13 @@ DhtRunner::loop_()
     if (nstatus4 != status4 || nstatus6 != status6) {
         status4 = nstatus4;
         status6 = nstatus6;
-        if (statusCb)
-            statusCb(status4, status6);
+        if (!statusCbs.empty())
```
Comment (aberaud) [inline src/dhtrunner.cpp:699]: if not needed

File: src/dhtrunner.cpp [hunk 2]
```diff
@@ -690,8 +696,13 @@ DhtRunner::loop_()
...
+        {
+            for (auto& cb : statusCbs){
+                if (cb)
```
Comment (aberaud) [inline src/dhtrunner.cpp:702]: check not needed

File: src/dhtrunner.cpp [hunk 3]
```diff
@@ -180,9 +180,8 @@ DhtRunner::run(const Config& config, Context&& context)
         throw;
     }
 
-    if (context.statusChangedCallback) {
-        statusCb = std::move(context.statusChangedCallback);
-    }
+    statusCbs.clear();
+    statusCbs.emplace_back(std::move(context.statusChangedCallback));
```
Comment (aberaud) [inline src/dhtrunner.cpp:184]: only emplace if `statusChangedCallback` is not `{}`.

File: include/opendht/dhtrunner.h [hunk 4]
```diff
@@ -519,7 +519,8 @@ class OPENDHT_PUBLIC DhtRunner {
 
     NodeStatus status4 {NodeStatus::Disconnected},
                status6 {NodeStatus::Disconnected};
-    StatusCallback statusCb {nullptr};
+
+    std::list<StatusCallback> statusCbs {};
```
Comment (aberaud) [inline include/opendht/dhtrunner.h:523]: could be a vector

File: include/opendht/dhtrunner.h [hunk 5]
```diff
@@ -409,7 +409,7 @@ class OPENDHT_PUBLIC DhtRunner {
     void run(const Config& config, Context&& context);
 
     void setOnStatusChanged(StatusCallback&& cb) {
-        statusCb = std::move(cb);
+        statusCbs.emplace_back(std::move(cb));
```
Comment (aberaud) [inline include/opendht/dhtrunner.h:412]: emplace if not empty

File: src/dhtrunner.cpp [hunk 6]
```diff
@@ -264,6 +263,13 @@ DhtRunner::run(const Config& config, Context&& context)
                 }
             }
         }
+        if (config.peer_discovery && config.peer_publish) {
```
Comment (aberaud) [inline src/dhtrunner.cpp:266]: what if only `peer_discovery` or `peer_publish` is true ?

**Review by aberaud — CHANGES_REQUESTED** [review]

File: src/dhtrunner.cpp (hunk 6 above)
Comment (AmnaSnene) [reply src/dhtrunner.cpp:266]: It should be `config.peer_discovery or config.peer_publish` so i changed it to `if (peerDiscovery_)`.

**Review by aberaud — APPROVED** [review]

**Pull request merged**